### PR TITLE
OSDOCS#4981: Allow mirroring images by tags

### PR DESCRIPTION
**Pull request by mburke5678**
[Allow mirroring images by tags](https://issues.redhat.com/browse/OCPNODE-521)
[Check mirror registry for policy](https://issues.redhat.com/browse/OSDOCS-4981)

Previews:
[Image source](https://56250--docspreview.netlify.app/openshift-enterprise/latest/cicd/builds/creating-build-inputs.html#builds-image-source_creating-build-inputs) -- Updated the NOTE
[Configuring image registry repository mirroring](https://56250--docspreview.netlify.app/openshift-enterprise/latest/post_installation_configuration/preparing-for-users.html#images-configuration-registry-mirror_post-install-preparing-for-users) -- Updates
[Converting ImageContentSourcePolicy (ICSP) files for image registry repository mirroring](https://56250--docspreview.netlify.app/openshift-enterprise/latest/post_installation_configuration/preparing-for-users.html#images-configuration-registry-mirror-convert_post-install-preparing-for-users) -- New module
[Disabling the Machine Config Operator from automatically rebooting](https://56250--docspreview.netlify.app/openshift-enterprise/latest/support/troubleshooting/troubleshooting-operator-issues.html#troubleshooting-disabling-autoreboot-mco_troubleshooting-operator-issues) -- Updated the second bullet in NOTE
[About the Machine Config Operator](https://56250--docspreview.netlify.app/openshift-enterprise/latest/architecture/control-plane.html#about-machine-config-operator_control-plane) -- Updated the second bullet in IMPORTANT

Configuring image registry repository mirroring and Converting ImageContentSourcePolicy (ICSP) files for image registry repository mirroring in *Images -> [Image configuration resources](https://56250--docspreview.netlify.app/openshift-enterprise/latest/updating/updating-restricted-network-cluster/restricted-network-update.html)*

Configuring image registry repository mirroring and Converting ImageContentSourcePolicy (ICSP) files for image registry repository mirroring in *Updating -> Updating a disconnected environment -> [Updating disconnected environments without the OpenShift Update Service](https://56250--docspreview.netlify.app/openshift-enterprise/latest/updating/updating-restricted-network-cluster/restricted-network-update.html)*

### CODE REVIEW

File: modules/images-configuration-registry-mirror.adoc [hunk 1]
```diff
@@ -66,53 +83,69 @@ In this example, you have a container image registry that is named `example.io`
...
+<7> Indicates the fallback policy if the image pull fails:
+** `AllowContactingSource` allows continued attempts to pull the image from the repository. This is the default. 
+** `NeverContactSource` prevents continued attempts to pull the image from the repository.
+<8> You can configure a namespace inside a registry to use any image in that namespace. If you use a registry domain as a source, the object is applied to all repositories from the registry.
+<9> If you configure the registry name, the object is applied to all repositories from a source registry to a mirror registry.
+<10> Pulls the image `mirror.example.net/image@sha256:...`.
+<11> Pulls the image `myimage` in the source registry namespace from the mirror `mirror.example.net/myimage@sha256:...`.
+<12> Pulls the image `registry.example.com/example/myimage` from the mirror registry `mirror.example.net/registry-example-com/example/myimage@sha256:...`. The `ImageContentSourcePolicy` resource is applied to all repositories from a source registry to a mirror registry `mirror.example.net/registry-example-com`.
```
Comment (mburke5678) [inline modules/images-configuration-registry-mirror.adoc:139]: @QiWang19 Can you PTAL at 10, 11, and 12. They don't seem correct to my reading. But I am not expert here.

**Comment by mburke5678** [comment]
@QiWang19 PTAL at this PR. Thank you!

File: modules/troubleshooting-disabling-autoreboot-mco.adoc [hunk 2]
```diff
@@ -17,7 +17,7 @@ The following modifications do not trigger a node reboot:
 ** Changes to the global pull secret or pull secret in the `openshift-config` namespace.
 ** Automatic rotation of the `/etc/kubernetes/kubelet-ca.crt` certificate authority (CA) by the Kubernetes API Server Operator.
 
-* When the MCO detects changes to the `/etc/containers/registries.conf` file, such as adding or editing an `ImageContentSourcePolicy` object, it drains the corresponding nodes, applies the changes, and uncordons the nodes. The node drain does not happen when there is a mirror configuration change in the `ImageContentSourcePolicy` (ICSP) object such as a new mirror added to an existing registry or a new registry added that has the setting `mirror-by-digest-only=true`.
+* When the MCO detects changes to the `/etc/containers/registries.conf` file, such as adding or editing an `ImageDigestMirrorSet` or `ImageTagMirrorSet` object, it drains the corresponding nodes, applies the changes, and uncordons the nodes. The node drain does not happen when there is a mirror configuration change in the object, such as adding a new mirror to an existing registry or adding a new registry that has the `mirrorSourcePolicy: AllowContactingSource` setting.
```
Comment (QiWang19) [inline modules/troubleshooting-disabling-autoreboot-mco.adoc:20]: adding a new mirror to an existing registry by ImageDigestMirrorSet

(Any change that introduced tag pull will still have node drain)

File: modules/understanding-machine-config-operator.adoc [hunk 3]
```diff
@@ -56,7 +56,7 @@ The following modifications do not trigger a node reboot:
 ** Changes to the global pull secret or pull secret in the `openshift-config` namespace.
 ** Automatic rotation of the `/etc/kubernetes/kubelet-ca.crt` certificate authority (CA) by the Kubernetes API Server Operator.
 
-* When the MCO detects changes to the `/etc/containers/registries.conf` file, such as adding or editing an `ImageContentSourcePolicy` object, it drains the corresponding nodes, applies the changes, and uncordons the nodes. The node drain does not happen when there is a mirror configuration change in the `ImageContentSourcePolicy` (ICSP) object such as a new mirror added to an existing registry or a new registry added that has the setting `mirror-by-digest-only=true`. 
+* When the MCO detects changes to the `/etc/containers/registries.conf` file, such as adding or editing an `ImageDigestMirrorSet` or `ImageTagMirrorSet` object, it drains the corresponding nodes, applies the changes, and uncordons the nodes. The node drain does not happen when there is a mirror configuration change in the object, such as adding a new mirror to an existing registry or adding a new registry that has the `mirrorSourcePolicy: AllowContactingSource` setting.
```
Comment (QiWang19) [inline modules/understanding-machine-config-operator.adoc:59]: point out ImageDigestMirrorSet here

**Review by QiWang19 — COMMENTED** [review]
Hi @mburke5678 , left some comments, thanks.
Will this remove the current document for ImageContentSourcePolicy? Just let you know the ImageContentSource will continue to function.

File: modules/images-configuration-registry-mirror.adoc [hunk 4]
```diff
@@ -40,14 +44,29 @@ requested from the source repository.
...
 ====
-You can only configure global pull secrets for clusters that have an `ImageContentSourcePolicy` object. You cannot add a pull secret to a project.
+If your cluster uses an `ImageDigestMirrorSet` or `ImageTagMirrorSet` object to configure repository mirroring, you can use only global pull secrets for mirrored registries. You cannot add a pull secret to a project.
 ====
 
 .Prerequisites
-* Access to the cluster as a user with the `cluster-admin` role.
+* Ensure you have access to the cluster as a user with the `cluster-admin` role.
+
+* Ensure there are no `ImageContentSourcePolicy` objects on your cluster. You can create new `ImageDigestMirrorSet` or `ImageTagMirrorSet` objects to replace any `ImageContentSourcePolicy` objects you delete.
```
[the remark this answers is not in the document]
Comment (mburke5678) [reply modules/images-configuration-registry-mirror.adoc:53]: @QiWang19 What is the command?

File: modules/images-configuration-registry-mirror.adoc [hunk 5]
```diff
@@ -66,53 +92,69 @@ In this example, you have a container image registry that is named `example.io`
...
-<5> If you configure the registry name, the `ImageContentSourcePolicy` resource is applied to all repositories from a source registry to a mirror registry.
-<6> Pulls the image `mirror.example.net/image@sha256:...`.
-<7> Pulls the image `myimage` in the source registry namespace from the mirror `mirror.example.net/myimage@sha256:...`.
-<8> Pulls the image `registry.example.com/example/myimage` from the mirror registry `mirror.example.net/registry-example-com/example/myimage@sha256:...`. The `ImageContentSourcePolicy` resource is applied to all repositories from a source registry to a mirror registry `mirror.example.net/registry-example-com`.
-
-. Create the new `ImageContentSourcePolicy` object:
+<1> Indicates the API to use with this CR. This must be `config.openshift.io/v1`.
+<2> Indicates the kind of object according to the pull type, either:
+** `ImageDigestMirrorSet`. Pulls a digest reference image.
+** `ImageTagMirrorSet`. Pulls a tag-based image.
```
Comment (QiWang19) [inline modules/images-configuration-registry-mirror.adoc:134]:
```suggestion
** `ImageTagMirrorSet`. Pulls a tag reference image.
```

**Review by QiWang19 — COMMENTED** [review]
@mburke5678 Thanks.

The changes LGTM other than one nit.

**Comment by mburke5678** [comment]
@lyman9966 Can you PTAL at my PR? Thank you very much in advance.

File: modules/images-configuration-registry-mirror.adoc [hunk 6]
```diff
@@ -66,53 +92,69 @@ In this example, you have a container image registry that is named `example.io`
...
 +
 [source,terminal]
 ----
 $ oc create -f registryrepomirror.yaml
 ----
 +
-After the `ImageContentSourcePolicy` object is created, the new settings are deployed to each node and the cluster starts using the mirrored repository for requests to the source repository.
+After the object is created, the Machine Config Operator (MCO) cordons the nodes as the new settings are deployed to each node. The MCO restarts the nodes for an `ImageTagMirrorSet` object only. The MCO does not restart the nodes for `ImageDigestMirrorSet` objects. When the nodes are uncordoned, the cluster starts using the mirrored repository for requests to the source repository.
 
 . To check that the mirrored configuration settings, are applied, do the following on one of the nodes.
```
Comment (lyman9966) [inline modules/images-configuration-registry-mirror.adoc:159]: please remove the first comma: 
To check that the mirrored configuration settings are applied, do the following on one of the nodes.

File: modules/images-configuration-registry-mirror.adoc [hunk 7]
```diff
@@ -66,53 +92,69 @@ In this example, you have a container image registry that is named `example.io`
...
+** `imageDigestMirrors` for an `ImageDigestMirrorSet` CR.
+** `imageTagMirrors` for an `ImageTagMirrorSet` CR. 
+<4> Indicates the name of the mirrored image registry and repository.
+<5> Optional. Indicates a secondary mirror repository for each target repository. If one mirror is down, the target repository can use another mirror.
+<6> Indicates the registry and repository source, which is the repository that is referred to in image pull specifications.
+<7> Optional. Indicates the fallback policy if the image pull fails:
+** `AllowContactingSource` allows continued attempts to pull the image from the source repository. This is the default. 
+** `NeverContactSource` prevents continued attempts to pull the image from the source repository.
+<8> Optional. Indicates a namespace inside a registry, which allows you to use any image in that namespace. If you use a registry domain as a source, the object is applied to all repositories from the registry.
+<9> Optional. Indicates a registry, which allows you to use any image in that namespace. If you specify a registry name, the object is applied to all repositories from a source registry to a mirror registry.
```
[the remark this answers is not in the document]
Comment (mburke5678) [reply modules/images-configuration-registry-mirror.adoc:145]: @QiWang19 Can you help with this question?

[the remark this answers is not in the document]
Comment (QiWang19) [reply modules/images-configuration-registry-mirror.adoc:145]: @mburke5678 The suggested change explains the difference between 8 and 9 more specifically that makes sense to me.

[the remark this answers is not in the document]
Comment (mburke5678) [reply modules/images-configuration-registry-mirror.adoc:145]: Yes it does. Thank you @QiWang19

**Review by kcarmichael08 — COMMENTED** [review]
I had some suggestions and questions; overall looks very good. Nice job with integrating this new content!

**Pull request merged**